### PR TITLE
Fix build on ESP32C3 LCD Kit by upgrading HAL

### PR DESCRIPTION
**Pull request by ottobonn**
Thanks for making this project; it's a major help to people like me trying to get into Rust programming on ESP32 while the documentation is still lacking.

I was not able to compile this code for the ESP32C3 LCD Kit due to several different errors, the most pressing of which was the fact that newer versions of the mipidsi crate require the DelayNs trait to be present on Delay, but older versions of the esp32c3-hal don't implement it. 

Upgrading the HAL in various places fixes the issue for this board. I haven't addressed the other boards in this PR; this is just to show what might be required to upgrade them. I did the corresponding upgrade to the DMA display driver as well.

### CODE REVIEW

File: esp32-c3-lcdkit/Cargo.toml [hunk 1]
```diff
@@ -6,7 +6,7 @@ edition = "2021"
 license = "MIT"
 
 [target.riscv32imc-unknown-none-elf.dependencies]
-hal = { package = "esp32c3-hal", version = "0.14.0" }
+hal = { package = "esp-hal", version = "0.16", features = ["esp32c3", "eh1"] }
```
Comment (ottobonn) [inline esp32-c3-lcdkit/Cargo.toml:9]: `eh1` is required for `DelayNs`

File: esp32-c3-lcdkit/src/app.rs
Comment (ottobonn) [inline esp32-c3-lcdkit/src/app.rs:1]: This file was unused

File: esp32-c3-lcdkit/src/main.rs [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #[global_allocator]
 static ALLOCATOR: esp_alloc::EspHeap = esp_alloc::EspHeap::empty();
 
+// use display_interface_spi::SPIInterfaceNoCS;
```
Comment (ottobonn) [inline esp32-c3-lcdkit/src/main.rs]:
```suggestion
```

Oops, typo

File: esp32-c3-lcdkit/src/main.rs [hunk 3]
```diff
@@ -140,56 +134,33 @@ fn main() -> ! {
...
-        .with_color_order(mipidsi::ColorOrder::Bgr)
-        .with_invert_colors(mipidsi::ColorInversion::Inverted)
-        .init(&mut delay, Some(lcd_reset))
-    {
-        Ok(disp) => disp,
-        Err(_) => {
-            panic!()
-        }
-    };
+    let mut display = mipidsi::Builder::new(mipidsi::models::GC9A01, di)
```
Comment (ottobonn) [inline esp32-c3-lcdkit/src/main.rs:158]: This API has changed quite a bit

File: spi-dma-displayinterface/Cargo.toml [hunk 4]
```diff
@@ -6,24 +6,17 @@ edition = "2021"
...
-esp32c3 = [ "esp32c3-hal" ]
-esp32c6 = [ "esp32c6-hal" ]
-esp32s2 = [ "esp32s2-hal" ]
-esp32s3 = [ "esp32s3-hal" ]
-
+esp32 = ["esp-hal/esp32"]
+esp32c3 = ["esp-hal/esp32c3"]
+esp32c6 = ["esp-hal/esp32c6"]
+esp32s2 = ["esp-hal/esp32s2"]
+esp32s3 = ["esp-hal/esp32s3"]
```
Comment (ottobonn) [inline spi-dma-displayinterface/Cargo.toml:22]: Different boards are now crate features of the HAL

**Comment by georgik** [comment]
@ottobonn Thank you very much for the upgrade. Amazing work!

I was waiting with upgrade until guys finished unification of HAL.
Let me do some tests and validation before merging PR.

**Comment by ottobonn** [comment]
The changes I made to the DMA display driver definitely broke the other boards, so this PR isn't complete yet!

**Comment by georgik** [comment]
@ottobonn I've tested the build and flash with the latest Rust. It works nicely. I've just noticed, that the project is downloading all hals, which seems to be reduntant. Other than that, the changes seems to be in the good direction.

Do you plan to update rest of boards? Just let me know, I can replicate your changes and test it with real HW for all combinations.

**Comment by ottobonn** [comment]
I don't think I quite know enough about the other boards to make the changes for them, so you can go ahead! Thanks for taking a look!

**Comment by georgik** [comment]
@bjoernQ Do you have any suggestion how to update esp-display-interface-spi-dma to esp-hal 0.17?

I've got half way through the compilation, but it seems that there is need for one additional generic. I would appreciate any hint. Here's PR for esp-display-interface-spi-dma: https://github.com/georgik/esp-display-interface-spi-dma/pull/1/files

**Comment by bjoernQ** [comment]
> @bjoernQ Do you have any suggestion how to update esp-display-interface-spi-dma to esp-hal 0.17?
> 
> I've got half way through the compilation, but it seems that there is need for one additional generic. I would appreciate any hint. Here's PR for esp-display-interface-spi-dma: https://github.com/georgik/esp-display-interface-spi-dma/pull/1/files

You probably don't need to be generic over `DmaMode` and just use `esp_hal::Blocking` I guess

**Comment by georgik** [comment]
During the upgrade of esp-display-interface-spi-dma we discovered a limitation of new HAL API https://github.com/esp-rs/esp-hal/issues/1512 . We need to figure out a proper solution to make the upgrade possible.

**Comment by jamessizeland** [comment]
I'm interested in trying out this example, is there anything I can do to help complete this PR?

**Comment by bjoernQ** [comment]
@georgik during the pre-release test for the current esp-hal release I updated https://github.com/bjoernQ/rust-esp32s3-ili9341 - that might help

**Comment by georgik** [comment]
Thank you @bjoernQ I'll try it.

**Comment by jamessizeland** [comment]
I just got https://github.com/bjoernQ/rust-esp32s3-ili9341 working on the esp32c3-lcd-kit with minimal changes 😊 on stable esp_hal 0.22

**Comment by georgik** [comment]
@jamessizeland Thank you for the notofication about ili. 
I will proceed with updating HAL and many dependencies on new branch [feature/upgrade-hal-2024-11](https://github.com/georgik/esp32-spooky-maze-game/tree/feature/upgrade-hal-2024-11)

I will begin with changes suggested by @ottobonn .

**Pull request merged**